include/watchdog.h

```diff
@@ -55,67 +71,242 @@ struct wdt_config {
+ */
+#define WDT_OPT_PAUSE_HALTED_BY_DBG	BIT(1)
+
+/**
+ * @brief Watchdog reset flag bit field mask shift.
+ */
+#define WDT_FLAG_RESET_SHIFT		(0)
+/**
+ * @brief Watchdog reset flag bit field mask.
+ */
+#define WDT_FLAG_RESET_MASK		(0x3 << WDT_FLAG_RESET_SHIFT)
+
+/**
+ * @name Watchdog Reset Behavior.
+ * Reset behavior after timeout.
+ * @{
+ */
+/** No reset */
+#define WDT_FLAG_RESET_NONE		(0 << WDT_FLAG_RESET_SHIFT)
+/** CPU core reset */
+#define WDT_FLAG_RESET_CPU_CORE		(1 << WDT_FLAG_RESET_SHIFT)
+/** Global SoC reset */
+#define WDT_FLAG_RESET_SOC		(2 << WDT_FLAG_RESET_SHIFT)
+/**
+ * @}
+ */
+
+/**
+ * @brief Watchdog timeout window.
+ *
+ * Each installed timeout needs feeding within the specified time window,
+ * otherwise the watchdog will trigger. If the watchdog instance does not
+ * support window timeouts then min value must be equal to 0.
+ *
+ * @param min Lower limit of watchdog feed timeout in milliseconds.
+ * @param max Upper limit of watchdog feed timeout in milliseconds.
+ *
+ * @note If specified values can not be precisely set they are always
+ *	 rounded up.
+ */
+struct wdt_window {
+	u32_t min;
+	u32_t max;
+};
+
+/** Watchdog callback. */
+typedef void (*wdt_callback_t)(struct device *dev, void *handle);
 
-typedef void (*wdt_api_enable)(struct device *dev);
-typedef void (*wdt_api_disable)(struct device *dev);
-typedef int (*wdt_api_set_config)(struct device *dev,
-				  struct wdt_config *config);
-typedef void (*wdt_api_get_config)(struct device *dev,
-				   struct wdt_config *config);
-typedef void (*wdt_api_reload)(struct device *dev);
+/**
+ * @brief Watchdog timeout configuration struct.
+ *
+ * @param window Timing parameters of watchdog timeout.
+ * @param handle Unique handle used to feed specific watchdog timeout.
+ *		 The same handle may be used for multistage timeouts.
+ *		 This handle is filled by the user, and will be supplied to
+ *		 callback. A good example is a handle pointing to the id of the
+ *		 thread that is responsible for feeding the corresponding
+ *		 timeout.
+ * @param callback Timeout callback. Passing NULL means that no callback
+ *		   will be run.
+ * @param next Pointer to the next timeout configuration. This pointer is used
+ *	       for watchdogs with staged timeouts functionality. Value must be
+ *	       NULL for single stage timeout.
+ * @param flags Bit field with following parts:
+ *
+ *     reset        [ 0 : 1 ]   - perform specified reset after timeout/callback
+ */
+struct wdt_timeout_cfg {
+	struct wdt_window window;
+	void *handle;
+	wdt_callback_t callback;
+#ifdef CONFIG_WDT_MULTISTAGE
+	struct wdt_timeout_cfg *next;
+#endif
+	u8_t flags;
+};
 
+/** @cond INTERNAL_HIDDEN */
 struct wdt_driver_api {
-	wdt_api_enable enable;
-	wdt_api_disable disable;
-	wdt_api_get_config get_config;
-	wdt_api_set_config set_config;
-	wdt_api_reload reload;
+	int (*wdt_api_setup)(struct device *dev, u8_t options);
```

Comment:
Can we remove the `wdt_api_` prefix from the member names? The prefix is fine if you define typedefs for the function prototypes, but it's redundant to also use the prefix in the struct members. See spi.h for an example.

Reply:
That's good question. I'd like to keep it consistent in some way with other drivers.
If we use spi as an example then I will prepare typedef for each function and pass it to this structure. It seems to be more readable. Please let me know what do You think about it

Reply:
It looks good